README.md
Quick links: [ [For Developers](#for-developers) | [Javadocs](http://bencvt.github.com/LibShapeDraw/apidocs/index.html) | [Demos](https://github.com/bencvt/LibShapeDraw/tree/master/projects/demos/src/main/java/libshapedraw/demos) | [Downloads](https://github.com/bencvt/LibShapeDraw/downloads) ]

# For Players

LibShapeDraw is a Minecraft client mod that is required by other mods.
It doesn't do anything on its own. Rather, it provides a set of flexible and
powerful drawing and animation tools for those other mods to use.

See the [official LibShapeDraw thread on minecraftforum.net](http://www.minecraftforum.net/topic/1458931-libshapedraw/)
for some screenshots and videos of what sort visual effects are possible.

## Installation

First of all, make sure that either
[ModLoader](http://www.minecraftforum.net/topic/75440-modloader/) or
[Forge](http://www.minecraftforge.net/forum/) is installed. LibShapeDraw is
compatible with either.

Next, download the jar and do one of the following, whichever you prefer:

 +  **Use a launcher utility to add the jar (recommended).** There are many
    launchers out there, e.g.
    [Magic Launcher](http://www.minecraftforum.net/topic/939149-magiclauncher/).

OR

 +  **Move the jar file to the `coremods` subdirectory.** This subdirectory can
    be found in your
    [Minecraft directory](http://www.minecraftwiki.net/wiki/.minecraft).
    If you use ModLoader rather than Forge, use the `mods` subdirectory instead.

OR

 +  **Manually patch the contents of the jar file into your `minecraft.jar`.**
    Old-school modding still works. Be sure to remove the `META-INF` folder.

## Compatibility

LibShapeDraw was designed with compatibility in mind. It does not modify *any*
vanilla classes directly and therefore should be compatible with virtually every
mod that works with ModLoader or Forge.

## Troubleshooting

If Minecraft is crashing, check [Minecaft Game Client Support](http://www.minecraftforum.net/forum/151-minecraft-game-client-support/)
on minecraftforums.net as a first step.

If it's not, the next step is to verify that LibShapeDraw is installed properly.
Browse to your Minecraft directory and look for a `mods/LibShapeDraw`
subdirectory. There should be a file named `LibShapeDraw.log` in it; you can
open it in a text editor. If this file doesn't exist or has an old timestamp
then LibShapeDraw is not installed correctly. Try reinstalling.

If you're still stuck with a specific problem, see the contact section at the
bottom. Please have the crash report and `LibShapeDraw.log` handy.

----

# For Developers

If your mod could use any of the following things, LibShapeDraw is the library
for you:

 +  The ability to create arbitrary shapes and **draw in the game world itself**
    *without* having to mess with OpenGL contexts, finding a render hook, or
    other tedious details.

 +  The ability to **smoothly animate** shapes (and anything else, really) using
    the easy-to-use Trident animation library built-in to LibShapeDraw.

 +  Vector3 and Color classes, full of well-documented and throughly-tested
    **convenience methods**. Want to calculate the pitch and yaw angles between
    two 3-dimensional points in a couple lines of code? No problem.

## Design goals

LibShapeDraw is designed to be easy to use, both for devs and for players. These
are the design goals to that end:

 +  **Minimal dependencies.** Either ModLoader or Forge is required to get up
    and running. That's all.

 +  **Maximal compatibility.** LibShapeDraw does not modify the bytecode of
    *any* vanilla Minecraft class. You are free to modify Minecraft classes in
    your own mod if needed; LibShapeDraw will not interfere.

 +  **Unobtrusive.** Pick and choose the components you want to use.
    LibShapeDraw is a toolkit for your mod to use. It is *not* a heavy
    DoEverythingThisWay framework.

 +  **Powerful.** What good is an API that doesn't let you do cool stuff? Check
    the demos for some of the many possibilities.

 +  **Concise and clear.** Convenience methods, fluent interfaces, etc. let you
    write less code to do more. That's what LibShapeDraw is all about.

 +  **Well-documented.** The key to success for any API, really.

 +  **Throughly tested.** A full JUnit test suite to help prevent bugs.

 +  **Open source.** MIT-licensed and open to community feedback and patches.

Enough high-level overview stuff. Let's dive in and actually use the API:

## Basic usage

Add the jar to your project's classpath and instantiate `LibShapeDraw` somewhere
in your code. Create and add some `Shape`s for the API instance to render.

Each Shape type is defined using classes like `Color`, `Vector3`, and
`ShapeTransform`. You can easily animate a shape by calling `animateStart` and
`animateStop` on these classes.

Plenty of sample code is available; see the *Documentation* section below.
Here's a quick example:

        // Create a 1x1x1 wireframe box at x=0, y=63 (sea level), z=0.
        this.libShapeDraw = new LibShapeDraw();
        Vector3 pointA = new Vector3(0, 63, 0);
        Vector3 pointB = new Vector3(1, 64, 1);
        WireframeCuboid shape = new WireframeCuboid(pointA, pointB);
        Color lineColor = Color.CYAN.copy();
        shape.setLineStyle(lineColor, 2.0F, false);
        libShapeDraw.addShape(shape);
        
        // Reposition it up a dozen blocks and make it 2x1x1 instead of 1x1x1.
        pointA.addY(12.0);
        pointB.addY(12.0).addX(1.0);
        
        // Animate it! Smoothly change color from cyan to green and back every 5
        // seconds, and also spin in place, rotating once every 7 seconds.
        lineColor.animateStartLoop(Color.GREEN, true, 5000);
        ShapeRotate rotate = new ShapeRotate(0.0, Axis.Y);
        shape.addTransform(rotate);
        rotate.animateStartLoop(360.0, false, 7000);

## How to add the LibShapeDraw jar to the classpath in MCP

[Minecraft Coder Pack (MCP)](http://mcp.ocean-labs.de/index.php/MCP_Releases)
is an excellent tool for creating mods, letting you work with deobfuscated
Minecraft code. Using external libraries (such as LibShapeDraw) involves a few
extra steps:

1.  **Get a copy of a LibShapeDraw release jar.** You can use either the normal
    release or the special dev release (named `LibShapeDraw-VERSION-dev.jar`).
    Check the [downloads list](https://github.com/bencvt/LibShapeDraw/downloads)
    for the latest.
    
    The dev release is recommended: either release type will work for
    *compiling* your mod, but the dev release also lets you *test/debug* your
    MCP mod without doing a full reobfuscate/deploy.

2.  **Copy the LibShapeDraw jar to `jars/bin`.**

3.  **Edit `conf/mcp.cfg`.** Scroll down to the `[RECOMPILE]` section. There is
    a property named `ClassPathClient`; this is where we will add a reference to
    the LibShapeDraw jar. Add `,%(DirJars)s/bin/LibShapeDraw-VERSION-dev.jar`
    (changing `VERSION` as appropriate) to the end of the line.

4.  **If you are using ModLoader, also copy the LibShapeDraw jar to
    `jars/mods`.** Skip this step if you're using Forge.

5.  **If you are using Eclipse, add the LibShapeDraw jar to the project's build
    build path.** Go to the Project Explorer pane and expand `jars/bin`.
    Right-click the LibShapeDraw jar, Build Path, Add to Build Path.
    
    The dev release also includes the source code for convenience, giving you
    easy access to the Javadocs in your IDE.

## Documentation

 +  [Javadocs](http://bencvt.github.com/LibShapeDraw/apidocs/index.html) are
    available.

 +  [Browse the demos](https://github.com/bencvt/LibShapeDraw/tree/master/projects/demos/src/main/java/libshapedraw/demos),
    located in `projects/demos`. To see the demos in action, you can
    [download](https://github.com/bencvt/LibShapeDraw/downloads) the pre-built
    demos jar and install it like any other mod.

 +  See [README-Trident.md](https://github.com/bencvt/LibShapeDraw/blob/master/README-Trident.md)
    for information about the built-in Trident animation library.

 +  See [README-contributing.md](https://github.com/bencvt/LibShapeDraw/blob/master/README-contributing.md)
    for information about contributing to LibShapeDraw itself, including build
    instructions and a list of features planned for future releases.

 +  If you'd like additional guidance, check the contacts section below.

## Contact

This project's official GitHub repo is located at
[github.com/bencvt/LibShapeDraw](https://github.com/bencvt/LibShapeDraw).
Anyone is free to open an issue.

You can also try the [official LibShapeDraw thread on minecraftforum.net](http://www.minecraftforum.net/topic/1458931-libshapedraw/).

Finally, feel free to ping me (bencvt) in IRC. I'm often lurking in various
channels related to Minecraft development on esper.net and freenode.net.


CHANGELOG.md
## implemented but not yet released
- more granular profiling
- added `ApiInfo.getModDirectory` and `LibShapeDraw.debugDump`

## 1.3.1
- added compatibility layer to support other mods that proxy
  `Minecraft.mcProfiler`, such as LiteLoader
- implement FML coremod interface to make installation more flexible
- added `AnimatedValue` convenience class
- added methods to make wireframe `GLUShape`s with arbitrary line widths
- primitive types are now serializable

## 1.3
- updated for Minecraft 1.4.5
- improved Trident animation library integration: added animation convenience
  methods to `Vector3` and defined the `Animates` interface
- primitive types are now directly comparable: added `equals`, `hashCode`, and
  `Vector3.equalsExact` methods
- added `Axis` enum and associated convenience methods to `Vector3` and
  `ShapeRotate`
- defined `XrayShape` interface
- added `Shape.onAdd`, `onRemove`, `onPreRender`, and `onPostRender` methods
- `LibShapeDraw.getShapes`, `getEventListeners`, and `Shape.getTransforms` now
  return read-only views. All modifications must use `addShape`, `removeShape`,
  `clearShapes`, etc.
- many documentation improvements

## 1.2.2
- updated for Minecraft 1.4.4
- improved Trident animation library integration: added animation convenience
  methods to `Color` and `ShapeTransform`

## 1.2.1
- updated for Minecraft 1.4.2
- added a section to the profiler so users can see the performance impact of
  LibShapeDraw on the shift-F3 debug screen; accessible under
  `root.gameRenderer.level.LibShapeDraw`
- a few minor internal changes to better support deobfuscation
- added `Timeline.playLoop(boolean reverse)` convenience method

## 1.2
- updated for Minecraft 1.4
- added update check
- added a few more convenience methods for primitive types
- fixed minor bug with `isVisibleWhenHidingGui` logic

## 1.1
- use a new rendering hook, eliminating graphical glitches near water
- improved Trident animation library integration: built-in interpolators for
  types in `libshapedraw.primitive`
- added `getPartialTick` method to `LSDPreRenderEvent` and `MinecraftAccess`
- added many other convenience methods, especially for `Vector3`
- added `mcmod.info` files and a logo for improved ForgeModLoader integration
  (Forge is still supported but *not* required)

## 1.0
- initial release for Minecraft 1.3.2


src/demos/java/LSDDemoTridentBasic.java
import libshapedraw.LibShapeDraw;
import libshapedraw.animation.trident.Timeline;
import libshapedraw.animation.trident.Timeline.RepeatBehavior;
import libshapedraw.primitive.Color;
import libshapedraw.primitive.ReadonlyColor;
import libshapedraw.shape.WireframeCuboid;
import libshapedraw.transform.ShapeRotate;
import libshapedraw.transform.ShapeScale;

/**
 * Create some shapes and animate them using the built-in Trident animation
 * library. Refer to README-Trident.md for more information.
 * <p>
 * Using Trident is the recommended method for animating Shapes, but it's not
 * the only way. You could, for example, manually update Vector3s/Colors/
 * ShapeTransforms in the LSDEventListener.onPreRender method.
 * <p>
 * Using Trident is worthwhile though: the end result is a lot less boilerplate
 * code for you to write. Also, Trident uses threading internally for better
 * performance.
 */
public class LSDDemoTridentBasic extends BaseMod {
    public static final String ABOUT =
            "Animate some shapes using the Trident animation library.\n" +
                    "Go to x=0, z=0 to see the shapes in action!";

    protected LibShapeDraw libShapeDraw = new LibShapeDraw();

    @Override
    public String getVersion() {
        return "demo";
    }

    @Override
    public void load() {
        libShapeDraw.verifyInitialized();
        // The Shapes will be created and their animations will start as soon
        // as ModLoader loads this mod. That means the Shapes are getting
        // updated even if we're on the main menu or on a loading screen.
        // 
        // This is a relatively harmless waste of a few CPU cycles. Trident
        // runs at 25 FPS, updating the Shapes at most every 40 milliseconds.
        // And in this case it's less than a dozen scalar values being updated.
        // 
        // If you're creating a lot of Shapes and associated Timelines
        // dynamically, a cleaner design will involve managing the Timelines
        // yourself. @see mod_LSDDemoTridentDynamic
        createBoxRotating();
        createBoxColorShifting();
        createBoxGrowing();
    }

    private void createBoxRotating() {
        WireframeCuboid box = new WireframeCuboid(8,63,0, 9,64,1);
        box.setLineStyle(Color.DODGER_BLUE.copy(), 3.0F, true);
        ShapeRotate rotate = new ShapeRotate(0.0, 0.0, 1.0, 0.0);
        box.addTransform(rotate);
        libShapeDraw.addShape(box);

        Timeline timeline = new Timeline(rotate);
        timeline.addPropertyToInterpolate("angle", 0.0, 360.0);
        timeline.setDuration(5000);
        timeline.playLoop(RepeatBehavior.LOOP);
    }

    private void createBoxColorShifting() {
        ReadonlyColor fromColor = Color.CRIMSON;
        ReadonlyColor toColor = Color.MEDIUM_BLUE.copy().setAlpha(0.2);

        WireframeCuboid box = new WireframeCuboid(8,63,2, 9,64,3);
        box.setLineStyle(fromColor.copy(), 3.0F, false);
        libShapeDraw.addShape(box);

        Timeline timeline = new Timeline(box.getLineStyle().getMainColor());
        timeline.addPropertyToInterpolate("red",   fromColor.getRed(),   toColor.getRed());
        timeline.addPropertyToInterpolate("blue",  fromColor.getBlue(),  toColor.getBlue());
        timeline.addPropertyToInterpolate("green", fromColor.getGreen(), toColor.getGreen());
        timeline.addPropertyToInterpolate("alpha", fromColor.getAlpha(), toColor.getAlpha());
        timeline.setDuration(750);
        timeline.playLoop(RepeatBehavior.REVERSE);
    }

    private void createBoxGrowing() {
        WireframeCuboid box = new WireframeCuboid(8,63,4, 9,64,5);
        box.setLineStyle(Color.GOLD.copy(), 3.0F, true);
        ShapeScale scale = new ShapeScale();
        box.addTransform(scale);
        libShapeDraw.addShape(box);

        Timeline timeline = new Timeline(scale.getScaleXYZ());
        timeline.addPropertyToInterpolate("x", 0.5F, 2.0F);
        timeline.addPropertyToInterpolate("z", 2.0F, 0.5F);
        timeline.setDuration(4500);
        timeline.playLoop(RepeatBehavior.REVERSE);

        // Since this is a simple WireframeCuboid, we could have interpolated on
        // getLowerCorner() and getUpperCorner() instead of using a ShapeScale
        // OpenGL transform. However it's generally better to let OpenGL handle
        // stuff whenever possible.
    }
}
